pkg/kubelet/cm/topologymanager/scope.go

```diff
@@ -109,10 +110,10 @@ func (s *scope) RemoveContainer(containerID string) error {
 	defer s.mutex.Unlock()
 
 	klog.InfoS("RemoveContainer", "containerID", containerID)
-	podUIDString := s.podMap[containerID]
-	delete(s.podMap, containerID)
-	if _, exists := s.podTopologyHints[podUIDString]; exists {
-		delete(s.podTopologyHints[podUIDString], containerID)
+	podUIDString, containerName, _ := s.podMap.GetContainerRef(containerID)
+	s.podMap.RemoveByContainerID(containerID)
+	if _, err := s.podMap.GetContainerID(podUIDString, containerName); err != nil {
+		delete(s.podTopologyHints[podUIDString], containerName)
 		if len(s.podTopologyHints[podUIDString]) == 0 {
 			delete(s.podTopologyHints, podUIDString)
 		}
```

Comment:
Some of this feels redundant. I think the following would be sufficient:
```
if podUIDString, containerName, exists := s.podMap.GetContainerRef(containerID); exists {
	s.podMap.RemoveByContainerID(containerID)
	delete(s.podTopologyHints[podUIDString], containerName)
	if len(s.podTopologyHints[podUIDString]) == 0 {
		delete(s.podTopologyHints, podUIDString)
	}
}
```

Reply:
It seems ok to `delete(s.podTopologyHints[podUIDString], containerName)` immediately after `s.podMap.RemoveByContainerID(containerID)` because we don't use it after the pod is created. But `podMap` is the mapping of (PodUid, ContainerName) to **ContainerID**, when a container restart (ContainerID changed), a new record will be added to `podMap`. Then the contents in `podMap` would be like `ContainerIDOld: {PodUID1, ContainerName1}, ContainerIDNew:{PodUID1, ContainerName1}`. If we delete `ContainerName1` in `podTopologyHints` when remove `ContainerIDOld` from `podMap`, it will cause the info between `podMap` and `podTopologyHints` is inconsistent. So I think it may be better to remove the container from `podTopologyHints` after deleting all records about this container in `podMap`.

Reply:
I see. You are handling the case where a container with the same `podUUID` and `containerName` is started with a different `ContainerID` (i.e. after a restart). We clearly don't want to delete that new containers topology hints inadvertently. 

I think this deserves a comment, since it is not obvious why we would need to re-call into the `podMap` in this way. How about:
```
// Get the podUID and containerName associated with the containerID to be removed and remove it
podUIDString, containerName, err := s.podMap.GetContainerRef(containerID)
if err != nil {
	return nil
}
s.podMap.RemoveByContainerID(containerID)

// In cases where a container has been restarted, it's possible that the same podUID and
// containerName are already associated with a *different* containerID now. Only remove
// the TopologyHints associated with that podUID and containerName if this is not true.
if _, err := s.podMap.GetContainerID(podUIDString, containerName); err != nil {
	delete(s.podTopologyHints[podUIDString], containerName)
	if len(s.podTopologyHints[podUIDString]) == 0 {
		delete(s.podTopologyHints, podUIDString)
	}
}
```

Reply:
Thanks a lot for your suggestion